Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,507 @@
+  ⟨v.toArray.set (v.size_eq.symm ▸ i) x, by {simp only [size_set]; exact v.size_eq}⟩
+
+/-- Set an element in a vector, or do nothing if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.-/
+def setD (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.setD i x, proof⟩
+  where
+    proof := by
+      simp only [size_setD]
+      exact v.size_eq
+      done
+/--
+Set an element in an array, or panic if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def set! (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.set! i x, proof⟩
+  where
+    proof := by
+      simp; exact v.size_eq
+      done
+
+
+/-- Appending a vector to another. -/
+def append {n m : Nat} : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+
+instance : HAppend (Vector α n) (Vector α m) (Vector α (n + m)) where
+  hAppend := append
+/--
+  `Vector.head` produces the head of a vector
+-/
+def head (v : Vector α (n+1)) :=
+  v.toArray.get ⟨0, v.size_eq.symm ▸ Nat.zero_lt_succ n⟩
+
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+
+def cons (x : α) (xs : Vector α n) : Vector α (n + 1) := Nat.add_comm 1 n ▸ (#v[x] ++ xs)
+
+/--Make a vector of size `n` with all cells containing `v`-/
+def mkVector (n : Nat) (v : α) : Vector α n :=
+  ⟨mkArray n v, proof⟩
+  where
+    proof := by
+      simp only [size_mkArray]
+      done
+
+/-- Create vector `#[]` -/
+def mkVector0 : Vector α 0 :=
+  mkEmpty 0
+
+/-- Create vector `#[v₁]` -/
+def mkVector1 (v₁ : α) : Vector α 1 :=
+  mkEmpty 1 |> push v₁
+
+/-- Create vector `#[v₁, v₂]` -/
+def mkVector2 (v₁ v₂ : α) : Vector α 2 :=
+  mkEmpty 2
+    |> push v₁
+    |> push v₂
+
+/-- Create vector `#[v₁, v₂, v₃]` -/
+def mkVector3 (v₁ v₂ v₃ : α) : Vector α 3:=
+  mkEmpty 3
+    |> push v₁
+    |> push v₂
+    |> push v₃
+
+
+/-- Create vector `#[v₁, v₂, v₃, v₄]` -/
+def mkVector4 (v₁ v₂ v₃ v₄ : α) : Vector α 4:=
+  mkEmpty 4
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅]` -/
+def mkVector5 (v₁ v₂ v₃ v₄ v₅ : α) : Vector α 5 :=
+  mkEmpty 5
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆]` -/
+def mkVector6 (v₁ v₂ v₃ v₄ v₅ v₆ : α) : Vector α 6 :=
+  mkEmpty 6
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆, v₇]` -/
+def mkVector7 (v₁ v₂ v₃ v₄ v₅ v₆ v₇ : α) : Vector α 7 :=
+  mkEmpty 7
+    |> push v₁
+    |> push v₂
+    |> push v₃
+    |> push v₄
+    |> push v₅
+    |> push v₆
+    |> push v₇
+
+/-- Create vector `#[v₁, v₂, v₃, v₄, v₅, v₆, v₇, v₈]` -/
+def mkVector8 (v₁ v₂ v₃ v₄ v₅ v₆ v₇ v₈ : α) : Vector α 8 :=
```

Comment:
I'm not sure we need `mkVector1` .. `mkVector8`. Then again, I'm also not sure why the Array API contains these. This needs some investigating...

Reply:
Perhaps @digama0 or @semorrison know more?

Reply:
I also fixed the notation in the docstring

Reply:
I think these are useful to quickly write expressions for short vectors in meta code. They would need unexpanders to work well. I think we could hold off on these until they are needed but let's see if Mario or Kim have thoughts.

Reply:
Makes sense. I suggest we leave this operation in place with a warning in the docstring.

Reply:
Found some uses for mkArray4 in [core](https://github.com/search?q=repo%3Aleanprover%2Flean4%20mkArray4&type=code)

Reply:
I think it makes sense to keep these definitions

Reply:
In that case, please add unexpanders as in core's Init.NotationExtra.

Reply:
Done. I reintroduced `mkVector0` to be consistent with the unexpanders for `Array`. I put them in a separate file called `Unexpanders.lean`

Reply:
I think `mkVector0` .. `mkVector8` could go in that file as well. That way they don't clutter Basic and it's more obvious what their purpose is.

Reply:
Done